### PR TITLE
feat: update relation generics (support Laravel >= 11.15)

### LINKED ISSUE #1284: `HasOneOrMany` wrongly uses `TRelatedModel` as the type for the declaring model of the relation
`HasOneOrMany` has only a single template parameter `TRelatedModel` which is bound to the model class which the relation points to. However, `HasOneOrMany` also uses `TRelatedModel` as the type for the model class which declares the relation.

https://github.com/nunomaduro/larastan/blob/ccdd599c464f9de6983a8d5e0f36daaf324b646f/stubs/HasOneOrMany.stub#L11-L20

The declaring model model (or parent or owning model) is not the same as the related model.

_Originally posted by @nagmat84 in https://github.com/nunomaduro/larastan/discussions/1275#discussioncomment-2933818_

### LINKED ISSUE #1550: Feature request: add support for detecting generic relation type for PendingHasThroughRelationships
- Larastan Version: 2.4.0
- `--level` used: 6

### Description
In https://github.com/laravel/framework/pull/45894/ a new `PendingHasThroughRelationship` was introduced to allow for easier through-relation building. 

However, when using these relations in my application, I was faced with the following error:

```
 ------ -------------------------------------------------------------------------------------------------------
  Line   app\Models\BaseModel.php
 ------ -------------------------------------------------------------------------------------------------------
  x      Parameter #1 $callback of method Illuminate\Database\Eloquent\PendingHasThroughRelationship::has()
         expects (callable(Illuminate\Database\Eloquent\Model):
         Illuminate\Database\Eloquent\Relations\HasMany|Illuminate\Database\Eloquent\Relations\HasOne)|string,
         Closure(App\Models\ThroughModel):
         Illuminate\Database\Eloquent\Relations\HasMany<App\Models\RelatedModel> given.
 ------ -------------------------------------------------------------------------------------------------------
```

### Laravel code where the issue was found
See also the PR mentioned above
```php
class Project extends Model
{
    public function deployments()
    {
        return $this->through($this->environments())
            ->has(fn (Environment $env) => $env->deployments());
    }

    public function environments()
    {
        return $this->hasMany(Environment::class);
    }
}

class Environment extends Model
{
    public function deployments()
    {
        return $this->hasMany(Deployment::class);
    }
}

class Deployment extends Model
{
    //
}
```

### LINKED ISSUE #1676: Method Builder::whereX() should return static(Builder<Model>) but returns Builder<Model>
- Larastan Version: 2.6.3
- `--level` used: 5
- Pull request with failing test: https://github.com/sanfair/larastan-playground/pull/1

### Description

When defining a new `whereX` method in custom builder with return type annotation phpstan shows errors.

In PR you can find both possible definitions that correctly execute at runtime, but trigger errors on static analysis.

Real application example: https://github.com/koel/koel/blob/master/app/Builders/AlbumBuilder.php#L10-L13
Note: in this case, an error can be resolved by specifying the return type as `self`, but it is not always correct.

Possible related issues from phpstan: https://github.com/phpstan/phpstan/issues/5132 https://github.com/phpstan/phpstan/issues/4648
Note: solution with `self` used from these issues

### Code where the issue was found

```php
use Illuminate\Database\Eloquent\Builder as BaseBuilder;

/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 *
 * @extends BaseBuilder<TModelClass>
 */
abstract class Builder extends BaseBuilder
{
    public function whereSomething($column, $values, $boolean = 'and'): static
    {
        return $this->whereNotIn(...func_get_args());
    }

    public function whereSomethingElse($column, $values, $boolean = 'and'): self
    {
        return $this->whereNotIn(...func_get_args());
    }
}
```

----

I managed to find the exact place where the type is defined. I tried to replace `\PHPStan\Type\Generic\GenericObjectType` with `\PHPStan\Type\StaticType` but then it fails in other places.

https://github.com/nunomaduro/larastan/blob/706784ec6b1bad487692730dc20aeee0ff1b19fa/src/Methods/EloquentBuilderForwardsCallsExtension.php#L150-L154

### PR DESCRIPTION
**Pull request by calebdw**
- [x] Added or updated tests
- [x] Documented user facing changes

Closes #1284, closes #1550, closes #1676

**Changes**
Hello!

This PR updates the Relation generics to align with https://github.com/laravel/framework/pull/51851 which implemented @nagmat84's fixes and notation from his PR (#1285) which introduced two new template types, `TDeclaringModel` and `TResult` (as well as `TIntermediateModel` for the *Through relations).

Here's an example of how Relations should now look:

```php
/** @return HasOne<Address, $this> */
public function address(): HasOne
{
    return $this->hasOne(Address::class);
}

/** @return HasMany<Post, $this> */
public function posts(): HasMany
{
    return $this->hasMany(Post::class);
}

/** @return BelongsToMany<Role, $this> */
public function roles(): BelongsToMany
{
    return $this->belongsToMany(Role::class);
}

/** @return HasOne<Mechanic, $this> */
public function mechanic(): HasOne
{
    return $this->hasOne(Mechanic::class);
}

/** @return HasOneThrough<Car, Mechanic, $this> */
public function car(): HasOneThrough
{
    return $this->hasOneThrough(Car::class, Mechanic::class);
}

/** @return HasManyThrough<Part, Mechanic, $this> */
public function parts(): HasManyThrough
{
    return $this->hasManyThrough(Part::class, Mechanic::class);
}
```

Of course Larastan can still parse the file to determine the generics if the phpdocs aren't provided, but it's always good to provide the generics.

There was a good bit of code I was able to clean up and greatly simplify along the way, as well as various bugs I encountered that needed to be fixed.

**Breaking changes**
I've tried my best to keep the breaking changes on 11.x, the older version still use just the `TRelatedModel` template (with the exception of the BelongsTo relation), but some changes were necessary to keep the code from becoming too complicated:
- renamed `TModelClass` to `TModel`
- renamed `TChildModel` to `TDeclaringModel`

### CODE REVIEW

File: src/Properties/ModelRelationsExtension.php
Comment (szepeviktor) [inline src/Properties/ModelRelationsExtension.php]:
```suggestion
                    // @todo remove this check when Larastan no longer supports < Laravel 11
```

**Comment by calebdw** [comment]
All tests are green :rocket:

File: composer.json
Comment (spawnia) [inline composer.json]:
```suggestion
        "test:types": "phpstan analyse --ansi --memory-limit=256M --verbose",
```

File: phpstan.neon.dist
Comment (spawnia) [inline phpstan.neon.dist:12]:
```suggestion
    # Supporting different Laravel versions requires ignoring errors that only occur in some cases
    reportUnmatchedIgnoredErrors: false
```

File: src/Properties/ModelRelationsExtension.php
Comment (spawnia) [inline src/Properties/ModelRelationsExtension.php]: Can you use `@phpstan-ignore` instead and perhaps provide a reason? I assume something like this:

```php
// @phpstan-ignore always.true (we support different Laravel versions)
```

File: src/Types/ModelRelationsDynamicMethodReturnTypeExtension.php
Comment (spawnia) [inline src/Types/ModelRelationsDynamicMethodReturnTypeExtension.php]: Again, this could be a bit more precise. Perhaps we can avoid this error altogether by declaring `LARAVEL_VERSION` a dynamic constant? See https://phpstan.org/config-reference#constants.

File: stubs/10.0.0/HasManyThrough.stub
Comment (spawnia) [inline stubs/10.0.0/HasManyThrough.stub:56]: Oh wow, that is super nice!

**Review by spawnia — COMMENTED** [review]
Very awesome, great effort! I installed your branch in a large project and got everything mostly working.

File: src/Types/ModelRelationsDynamicMethodReturnTypeExtension.php
Comment (calebdw) [reply src/Types/ModelRelationsDynamicMethodReturnTypeExtension.php]: Fixed!

File: src/Properties/ModelRelationsExtension.php
Comment (calebdw) [reply src/Properties/ModelRelationsExtension.php]: I tried that and it didn't work because the identifier changes depending on the Laravel version (e.g., `always.false).

However, I've reworked all of the `phpstan-ignore-[next-]line` tags to use the new syntax with comments on why they're being ignored.

File: stubs/10.0.0/HasManyThrough.stub
Comment (calebdw) [reply stubs/10.0.0/HasManyThrough.stub:56]: Yeah, it cleans up a lot of phpstan extensions that were used for this purpose!

**Review by spawnia — APPROVED** [review]
Works great, I have no futher nits to pick.

**Comment by pataar** [comment]
Any idea when this one is going through? 😁 
Great work btw on the generics! Both in the framework and here

**Comment by EliW** [comment]
Just chiming in alongside @pataar here ... any idea when this will be released?  We've now skipped 3 (bugfix level) versions of Laravel waiting for this to updated and work :)    Thanks again

**Comment by calebdw** [comment]
For those that are interested, I've gone ahead and published these changes on my own fork ([calebdw/larastan](https://github.com/calebdw/larastan)). Feel free to check it out while we wait for this to be merged!

**Comment by HDVinnie** [comment]
> For those that are interested, I've gone ahead and published these changes on my own fork ([calebdw/larastan](https://github.com/calebdw/larastan)). Feel free to check it out while we wait for this to be merged!

Is this something you plan to maintain? Appreciate all the work you have been doing and 100% willing to switch if its going to be maintained and not just dead once merged here.

**Comment by calebdw** [comment]
@HDVinnie, 

Yes I plan on maintaining it (for the foreseeable future at least). The goal is to just offer quicker access to more "bleeding edge" features (as PHPStan likes to call it) than is available in the upstream (i.e., this) package.

**Comment by EliW** [comment]
> @HDVinnie,
> 
> Yes I plan on maintaining it (for the foreseeable future at least). The goal is to just offer quicker access to more "bleeding edge" features (as PHPStan likes to call it) than is available in the upstream (i.e., this) package.

Thanks for maintaining this Caleb ... but honest question, not knowing the 'politics' of the larastan/larastan repo itself.   But this PR fixes a fundamental problem, since the Laravel 11.15 release, not having this has caused major issues (at least in our codebase), which in turn has lead to my team currently skipping Laravel 11.15 through now 11.19 ...

This PR is approved (or so it would seem) ... and fixes an outstanding issue from 11.15 on.    Is there a reason why it's not being merged and released so that people can update their Laravel?

I will look into switching to using your 'bleeding edge' version, but I'm struggling to understand why a 'needed fix' that is breaking compatibility with Laravel 11.15 and upwards (so not bleeding edge, it's required for compatibility with the last 4 minor revs of Laravel), that has been approved, isn't just being merged?

**Comment by calebdw** [comment]
@EliW 

> This PR is approved (or so it would seem) ... and fixes an outstanding issue from 11.15 on. Is there a reason why it's not being merged and released so that people can update their Laravel?

The only approval on this PR is from another contributor who can't merge this. The maintainers of Larastan are @canvural, @nunomaduro, and @szepeviktor. However, Viktor is just a "janitor" here so he doesn't merge PRs, and Nuno (from what I've seen) tends to have his hands full with other Laravel projects so he's not too active here. Which leaves Can as the primary maintainer that approves and merges PRs.

I haven't heard from Can recently, but he's likely busy with life / work / other things and hasn't had a chance to look at this yet (which is completely understandable). Can maintains this in his *free* time (emphasis on free) and thus it's hard to estimate when this will be merged---it could be tomorrow or it could be in a month or two or even longer.

I've offered to help out as a maintainer (#1912), but unfortunately Can was not looking to expand the team at that time.

**Comment by EliW** [comment]
Thanks @calebdw for explanation.  TOTALLY understand on the _free time_ volunteer aspect.  Was just curious to understand if there was any bigger reasons I wasn't seeing.

**Comment by SanderMuller** [comment]
LGTM @canvural, @nunomaduro, and @szepeviktor

**Comment by canvural** [comment]
Hi,

I'm back from my vacation. So I'll take a look at this as soon as possible. But currently I have close to 900 GitHub notifications 🥲 Also this is big PR, so it'll take some to review it. Thank you for understanding.

**Comment by calebdw** [comment]
No worries @canvural! Glad you had a vacation but sorry for all the notification to catch up on haha

Let me know if you have any questions or if there's anything I can do to help :+1:

**Comment by SanderMuller** [comment]
> Hi,
> 
> I'm back from my vacation. So I'll take a look at this as soon as possible. But currently I have close to 900 GitHub notifications 🥲 Also this is big PR, so it'll take some to review it. Thank you for understanding.

Very understandable! Thanks for all your efforts. My apologies that my previous message was not a very contributing one.

**Comment by canvural** [comment]
Hi,

Initially I wanted to get this into 3.0 only. But then I remember PHPStan 2.0 is coming out soon and we'd need to make another major version increase for that. So decided to include this in 2.x

Can you please fix the merge conflicts? then we can merge and make a release. Thank you

**Comment by calebdw** [comment]
@canvural, this has been rebased and all the tests are passing :+1:

Before you cut a new release, can we please try to get some of my other PRs merged? I've been stepping over myself in the different PRs, so I'll likely need to resolve conflicts in other PRs as they're merged but it would be great to get them out the door if you don't mind.

Thanks!

**Comment by AJenbo** [comment]
If I can somehow be of help I'm also happy to step in and lend a hand.

**Comment by canvural** [comment]
Hi,

I couldn't merge this as is. I was struggling for a few days, tweaking, trying different stuff. But finally found a good balance. I cherry picked some of your commits, I adjusted some, and it's working.

Thank you!

**Comment by calebdw** [comment]
@canvural, could you please be more descriptive about why you couldn't merge it as-is and what you had issues with? For what it's worth, this has been published on my fork and has been used in production code without any issues so it's not like this hasn't been tested.

I would've appreciated the opportunity to address any concerns / questions and explain my reasoning for doing things the way I did---there were some edge cases and bugs I discovered and fixed along the way that might not have made it into your cherry picks. I didn't do anything willy-nilly, there were specific reasons for everything in this PR.

Best,

**Comment by canvural** [comment]
> @canvural, could you please be more descriptive about why you couldn't merge it as-is and what you had issues with? For what it's worth, this has been published on my fork and has been used in production code without any issues so it's not like this hasn't been tested.
> 
> I would've appreciated the opportunity to address any concerns / questions and explain my reasoning for doing things the way I did---there were some edge cases and bugs I discovered and fixed along the way that might not have made it into your cherry picks. I didn't do anything willy-nilly, there were specific reasons for everything in this PR.
> 
> Best,

Yeah, sorry that I didn't write in details. I was just trying to push it forwards, and later didn't find time.

I couldn't merge it as is, because:
- it changed generic type names also for Laravel 9 and 10. This is a breaking change and can't do it in a minor version.
- it had lots of changes that were not really related to the actual issue it is fixing.

I didn't want to write those here and ask you to fix again, because this PR was already open more than 2 months and you really did a lot and didn't want to bother you again.

As I said I tried to use as much as I can from your changes. Hope this clears it up. And apologize again for being not very clear.

**Comment by calebdw** [comment]
@canvural, thanks for the response.

> it changed generic type names also for Laravel 9 and 10. This is a breaking change and can't do it in a minor version.

I understand this, if you had mentioned that this was a requirement then I could've quickly implemented a static class like you did, I just originally left this out because I wasn't sure if this would be acceptable or not.

> it had lots of changes that were not really related to the actual issue it is fixing.

Perhaps, but these were things that needed to be fixed (*still need* to be fixed given that these changes weren't merged). As I said, I came across various issues that I just decided to fix in this branch, otherwise I would've been tripping over myself if I were to try to create individual PRs.

> I didn't want to write those here and ask you to fix again, because this PR was already open more than 2 months and you really did a lot and didn't want to bother you again.

I appreciate the thought, but unfortunately the changes you disregarded are going to have to be fixed again...

Best,

**Comment by canvural** [comment]
> I appreciate the thought, but unfortunately the changes you disregarded are going to have to be fixed again...

If I see the issues, I can fix them.

**Pull request closed without merging**
